File: apps/ios/Sources/UI/Components/ConversationList.swift
// Zitrone — Copyright (C) 2026 Zitrone contributors
// Licensed under the GNU Affero General Public License v3.0 or later.
// See the LICENSE file in the repository root for full license text.
// SPDX-License-Identifier: AGPL-3.0-only

import SwiftUI

/// Conversation list per design_system.components.conversation_list_item.
/// CRITICAL: the preview text is ALWAYS "Encrypted message" — message content
/// is never previewed anywhere outside the open chat.
public struct ConversationList: View {
    public let conversations: [Conversation]
    public var onSelect: (Conversation) -> Void
    /// Long-press delete. Host must show an irreversible confirmation dialog
    /// before calling `MessageStore.deleteContact`.
    public var onDelete: ((Conversation) -> Void)?

    public init(conversations: [Conversation],
                onSelect: @escaping (Conversation) -> Void,
                onDelete: ((Conversation) -> Void)? = nil) {
        self.conversations = conversations
        self.onSelect = onSelect
        self.onDelete = onDelete
    }

    public var body: some View {
        ScrollView {
            LazyVStack(spacing: 0) {
                ForEach(conversations) { conversation in
                    Button {
                        onSelect(conversation)
                    } label: {
                        ConversationListItem(conversation: conversation)
                    }
                    .buttonStyle(ConversationRowStyle())
                    .contextMenu {
                        if let onDelete {
                            Button(role: .destructive) {
                                onDelete(conversation)
                            } label: {
                                Label("Delete contact", systemImage: "trash")
                            }
                        }
                    }
                    // Long-press affordance for users who don't discover the
                    // context menu: same action as the destructive menu item.
                    .onLongPressGesture(minimumDuration: 0.55) {
                        onDelete?(conversation)
                    }
                }
            }
        }
        // No opaque background of its own: the hosting screen paints the
        // ground AND the fingerprint watermark beneath this list — an opaque
        // fill here would hide the security paper behind every row (PR #8
        // review).
    }
}

public struct ConversationListItem: View {
    public let conversation: Conversation

    public init(conversation: Conversation) {
        self.conversation = conversation
    }

    public var body: some View {
        HStack(spacing: Spacing.s3) {
            LemonAvatar(name: conversation.contact.displayName,
                        verified: conversation.contact.verified)

            VStack(alignment: .leading, spacing: Spacing.s1) {
                Text(conversation.contact.displayName)
                    .font(SubFont.body(TypeScale.base, weight: .medium))
                    .foregroundColor(.textPrimary)
                    .lineLimit(1)
                // Never previews content — by design.
                Text("Encrypted message")
                    .font(SubFont.body(TypeScale.sm))
                    .foregroundColor(.textMuted)
            }

            Spacer()

            VStack(alignment: .trailing, spacing: Spacing.s1) {
                if let lastActivity = conversation.lastActivity {
                    Text(lastActivity, style: .time)
                        .font(SubFont.mono(TypeScale.xs))
                        .foregroundColor(.textMuted)
                }
                if conversation.unreadCount > 0 {
                    Text("\(conversation.unreadCount)")
                        .font(SubFont.body(TypeScale.xs, weight: .medium))
                        .foregroundColor(.textOnLemon)
                        .padding(.horizontal, Spacing.s2)
                        .padding(.vertical, 2)
                        .background(Capsule().fill(Color.lemon))
                }
            }
        }
        .padding(.horizontal, Spacing.s4)
        .padding(.vertical, Spacing.s3)
        .contentShape(Rectangle())
    }
}

/// 44pt circular avatar on a lemon gradient; 2pt green ring when the
/// contact's keys are verified.
public struct LemonAvatar: View {
    public let name: String
    public let verified: Bool
    public var size: CGFloat = 44

    public init(name: String, verified: Bool, size: CGFloat = 44) {
        self.name = name
        self.verified = verified
        self.size = size
    }

    public var body: some View {
        ZStack {
            Circle().fill(SubGradient.avatar)
            Text(initials)
                .font(SubFont.display(size * 0.4, weight: .semibold))
                .foregroundColor(.textOnLemon)
        }
        .frame(width: size, height: size)
        .overlay(
            Circle().strokeBorder(verified ? Color.verifiedGreen : .clear, lineWidth: 2)
        )
    }

    private var initials: String {
        let parts = name.split(separator: " ").prefix(2)
        let letters = parts.compactMap { $0.first.map(String.init) }
        return letters.isEmpty ? "?" : letters.joined().uppercased()
    }
}

private struct ConversationRowStyle: ButtonStyle {
    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .background(configuration.isPressed ? Color.backgroundElevated : .clear)
    }
}

File: apps/android/app/src/main/java/com/sublemonable/app/ui/components/ConversationList.kt
// Sublemonable — Copyright (C) 2026 Sublemonable contributors
// Licensed under the GNU Affero General Public License v3.0 or later.
// See the LICENSE file in the repository root for full license text.
// SPDX-License-Identifier: AGPL-3.0-only

package com.sublemonable.app.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.sublemonable.app.data.Conversation
import com.sublemonable.app.ui.theme.Lemon
import com.sublemonable.app.ui.theme.LemonZest
import com.sublemonable.app.ui.theme.MonoFamily
import com.sublemonable.app.ui.theme.PillShape
import com.sublemonable.app.ui.theme.TextMuted
import com.sublemonable.app.ui.theme.TextOnLemon
import com.sublemonable.app.ui.theme.TextPrimary
import com.sublemonable.app.ui.theme.TypeScale
import com.sublemonable.app.ui.theme.VerifiedGreen
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter

private val ListTimeFormatter: DateTimeFormatter =
    DateTimeFormatter.ofPattern("HH:mm").withZone(ZoneId.systemDefault())

/**
 * Conversation list (design_system.components.conversation_list_item).
 * Privacy rule baked in: the preview line is ALWAYS the literal text
 * "Encrypted message" — message content never appears outside an open chat.
 */
@Composable
fun ConversationList(
    conversations: List<Conversation>,
    onOpenConversation: (Conversation) -> Unit,
    modifier: Modifier = Modifier,
) {
    LazyColumn(modifier = modifier.fillMaxSize()) {
        items(conversations, key = { it.id }) { conversation ->
            ConversationListItem(
                conversation = conversation,
                onClick = { onOpenConversation(conversation) },
            )
        }
    }
}

@Composable
fun ConversationListItem(
    conversation: Conversation,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier = modifier
            .fillMaxWidth()
            .clickable(onClick = onClick)
            .padding(horizontal = 16.dp, vertical = 10.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        LemonAvatar(
            name = conversation.displayName,
            verified = conversation.verified,
        )

        Column(modifier = Modifier.weight(1f)) {
            Text(
                text = conversation.displayName,
                style = MaterialTheme.typography.titleMedium,
                color = TextPrimary,
                maxLines = 1,
            )
            // ALWAYS "Encrypted message" — never a content preview.
            Text(
                text = "Encrypted message",
                style = MaterialTheme.typography.bodySmall,
                color = TextMuted,
                maxLines = 1,
            )
        }

        Column(horizontalAlignment = Alignment.End, verticalArrangement = Arrangement.spacedBy(4.dp)) {
            if (conversation.lastActivityMs > 0L) {
                Text(
                    text = ListTimeFormatter.format(Instant.ofEpochMilli(conversation.lastActivityMs)),
                    fontFamily = MonoFamily,
                    fontSize = TypeScale.Xs,
                    color = TextMuted,
                )
            }
            if (conversation.unreadCount > 0) {
                Box(
                    modifier = Modifier
                        .background(Lemon, PillShape)
                        .padding(horizontal = 8.dp, vertical = 2.dp),
                ) {
                    Text(
                        text = conversation.unreadCount.toString(),
                        style = MaterialTheme.typography.labelMedium,
                        fontWeight = FontWeight.Medium,
                        color = TextOnLemon,
                    )
                }
            }
        }
    }
}

/** 44dp circular avatar on a lemon gradient; verified ring in green. */
@Composable
fun LemonAvatar(
    name: String,
    verified: Boolean,
    modifier: Modifier = Modifier,
) {
    Box(
        modifier = modifier
            .size(44.dp)
            .then(
                if (verified) {
                    Modifier.border(2.dp, VerifiedGreen, CircleShape)
                } else {
                    Modifier
                },
            )
            .padding(if (verified) 3.dp else 0.dp)
            .background(
                brush = Brush.linearGradient(listOf(Lemon, LemonZest)),
                shape = CircleShape,
            ),
        contentAlignment = Alignment.Center,
    ) {
        Text(
            text = name.firstOrNull()?.uppercase() ?: "?",
            style = MaterialTheme.typography.titleMedium,
            color = TextOnLemon,
        )
    }
}
